WordPress/gutenberg
# Navigation screen: Add opt-in Navigation block rendering

Implements the majority of https://github.com/WordPress/gutenberg/issues/24364. Continues on from the work started in https://github.com/WordPress/gutenberg/pull/22656.

![demo](https://user-images.githubusercontent.com/612155/90094828-5331bc00-dd72-11ea-946d-6899eaf7ef0f.gif)

Allows themes to opt-in to having `wp_nav_menu()` render a complete Navigation block in place of the menu. This allows arbitrarily complex Navigation block structures can be used in an existing theme.
    
When a theme declares `add_theme_support( 'block-nav-menus' )`, users are able to add non-link blocks to a Navigation using the Navigation screen.
    
This block tree is stored as menu items and re-assembled into a Navigation block for display using `render_block()` in the frontend.

If the user switches to a theme that does not support block menus, or disables this functionality, non-link blocks are no longer rendered in the frontend. Care is taken, however, to ensure that users can still see their data in `nav-menus.php`.

See https://github.com/WordPress/gutenberg/issues/24364 for more information on the approach.

### How to test

1. Note that in _Gutenberg > Navigation (beta)_, you can only add Link blocks to your Navigation. Note also that the frontend renders regular nav menu markup.

2. Patch your current theme to support block navigation. Here's how that looks for Twenty Twenty:
   
```diff
Index: src/wp-content/themes/twentytwenty/functions.php
===================================================================
--- src/wp-content/themes/twentytwenty/functions.php	(revision 48655)
+++ src/wp-content/themes/twentytwenty/functions.php	(working copy)
@@ -142,6 +142,7 @@
 	$loader = new TwentyTwenty_Script_Loader();
 	add_filter( 'script_loader_tag', array( $loader, 'filter_script_loader_tag' ), 10, 2 );
 
+	add_theme_support( 'block-nav-menus' );
 }
 
 add_action( 'after_setup_theme', 'twentytwenty_theme_support' );
Index: src/wp-content/themes/twentytwenty/header.php
===================================================================
--- src/wp-content/themes/twentytwenty/header.php	(revision 48655)
+++ src/wp-content/themes/twentytwenty/header.php	(working copy)
@@ -83,25 +83,25 @@
 
 					<?php
 					if ( has_nav_menu( 'primary' ) || ! has_nav_menu( 'expanded' ) ) {
-						?>
 
-							<nav class="primary-menu-wrapper" aria-label="<?php esc_attr_e( 'Horizontal', 'twentytwenty' ); ?>" role="navigation">
+						if ( has_nav_menu( 'primary' ) ) {
 
-								<ul class="primary-menu reset-list-style">
+							wp_nav_menu(
+								array(
+									'container'      => '',
+									'items_wrap'     => '%3$s',
+									'theme_location' => 'primary',
+								)
+							);
 
-								<?php
-								if ( has_nav_menu( 'primary' ) ) {
+						} elseif ( ! has_nav_menu( 'expanded' ) ) {
+							?>
 
-									wp_nav_menu(
-										array(
-											'container'  => '',
-											'items_wrap' => '%3$s',
-											'theme_location' => 'primary',
-										)
-									);
+							<nav class="primary-menu-wrapper" aria-label="<?php esc_attr_e( 'Horizontal', 'twentytwenty' ); ?>" role="navigation">
 
-								} elseif ( ! has_nav_menu( 'expanded' ) ) {
+								<ul class="primary-menu reset-list-style">
 
+									<?php
 									wp_list_pages(
 										array(
 											'match_menu_classes' => true,
@@ -110,15 +110,14 @@
 											'walker'   => new TwentyTwenty_Walker_Page(),
 										)
 									);
+									?>
 
-								}
-								?>
-
 								</ul>
 
 							</nav>
 
-						<?php
+							<?php
+						}
 					}
 
 					if ( true === $enable_header_search || has_nav_menu( 'expanded' ) ) {
```

2. The frontend should now render Navigation block markup.

3. Go back to _Gutenberg > Navigation (beta)_ and add a Search block. The Search block should now display on the frontend.

4. Switch to a different theme. The frontend should now display regular nav menu markup without a Search block.

5. Go back to _Gutenberg > Navigation (beta)_. The editor should now not allow a Search block to be inserted. The existing Search block should still be displayed.

6. Go to _Appearance > Menus_. A Gutenberg Block widget should appear containing the Search block.

### Follow up tasks

- Investigate making this setting _per menu location_. https://github.com/WordPress/gutenberg/pull/24503#issuecomment-673198300
- Add UI which indicates when a block is not going to be rendered on the frontend. https://github.com/WordPress/gutenberg/pull/24503#discussion_r471872542
- Investigate registering a different 'missing' block. https://github.com/WordPress/gutenberg/pull/24503#discussion_r471894496
